src/main/java/org/opentripplanner/routing/graphfinder/PlaceFinderTraverseVisitor.java

```diff
@@ -167,17 +177,49 @@ private boolean stopHasPatternsWithMode(RegularStop stop, Set<TransitMode> modes
       .anyMatch(modes::contains);
   }
 
+  private boolean stopIsIncludedByStopFilter(RegularStop stop) {
+    return filterByStops.isEmpty() || filterByStops.contains(stop.getId());
+  }
+
+  private boolean stopIsIncludedByStationFilter(RegularStop stop) {
+    return (
+      ((filterByStations.isEmpty() || filterByStations.contains(stop.getParentStation().getId())))
+    );
+  }
+
+  private boolean stopIsIncludedByModeFilter(RegularStop stop) {
+    return filterByModes.isEmpty() || stopHasPatternsWithMode(stop, filterByModes);
+  }
+
   private void handleStop(RegularStop stop, double distance) {
-    if (filterByStops != null && !filterByStops.contains(stop.getId())) {
-      return;
-    }
+    /*
+     Do not consider stop if it is not included in the stop or mode filter
+     or if it or its parent station has already been seen.
+     */
     if (
-      includeStops &&
-      !seenStops.contains(stop.getId()) &&
-      (filterByModes == null || stopHasPatternsWithMode(stop, filterByModes))
+      (includeStations && !stop.isPartOfStation() && !stopIsIncludedByStopFilter(stop)) ||
+      (!includeStations && !stopIsIncludedByStopFilter(stop)) ||
+      (stop.isPartOfStation() && !stopIsIncludedByStationFilter(stop)) ||
+      seenStops.contains(stop.getId()) ||
+      seenStops.contains(stop.getStationOrStopId()) ||
+      !stopIsIncludedByModeFilter(stop)
     ) {
-      placesFound.add(new PlaceAtDistance(stop, distance));
+      return;
+    }
+    if (includeStops && includeStations) {
+      if (stop.getParentStation() != null) {
+        seenStops.add(stop.getParentStation().getId());
+        placesFound.add(new PlaceAtDistance(stop.getParentStation(), distance));
+      } else {
+        seenStops.add(stop.getId());
+        placesFound.add(new PlaceAtDistance(stop, distance));
+      }
+    } else if (includeStations && stop.getParentStation() != null) {
+      seenStops.add(stop.getParentStation().getId());
+      placesFound.add(new PlaceAtDistance(stop.getParentStation(), distance));
+    } else if (includeStops) {
       seenStops.add(stop.getId());
+      placesFound.add(new PlaceAtDistance(stop, distance));
     }
```

Comment:
Isn't it enough to have just these two checks and remove the first if? If I'm not missing anything, I think this also covers the case where both `includeStops` and `includeStations` is true.